apm-agent-plugins/apm-rabbitmq-plugin/src/main/java/co/elastic/apm/agent/rabbitmq/RabbitMQProducerInstrumentation.java

```diff
@@ -0,0 +1,134 @@
+ */
+package co.elastic.apm.agent.rabbitmq;
+
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TextHeaderSetter;
+import com.rabbitmq.client.AMQP;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.classLoaderCanLoadClass;
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public class RabbitMQProducerInstrumentation extends RabbitMQBaseInstrumentation {
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("com.rabbitmq.client.Channel"));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("basicPublish")
+            .and(takesArguments(6));
+    }
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        return not(isBootstrapClassLoader()).and(classLoaderCanLoadClass("com.rabbitmq.client.Channel"));
+    }
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return RabbitProducerAdvice.class;
+    }
+
+    public static class RabbitProducerAdvice {
+
+        private RabbitProducerAdvice() {}
+
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        @Nullable
+        public static Object onBasicPublish(@Advice.Argument(0) final String exchange/*, @Advice.Argument(value = 4, readOnly = false) @Nullable AMQP.BasicProperties basicProperties*/) {
+            if (!tracer.isRunning() || tracer.getActive() == null) {
+                return null;
+            }
+
+            final AbstractSpan<?> activeSpan = tracer.getActive();
+            if (activeSpan == null) {
+                return null;
+            }
+
+            Span exitSpan = activeSpan.createExitSpan();
+            if (null == exitSpan) {
+                return null;
+            }
+
+            exitSpan.withType("messaging").withSubtype("rabbitmq")
+                .withAction("send")
+                .withName("RabbitMQ message sent to ")
+                .appendToName(exchange);
+
+            // TODO: Propagate trace context
+            //final TextHeaderSetter<HashMap<String, Object>> textHeaderSetter = new RabbitMQTextHeaderSetter();
+            //basicProperties = propagateTraceContext(exitSpan, basicProperties, textHeaderSetter);
+
+            /*
+            TODO: Transaction context
+            exitSpan.getContext().getMessage().withQueue(exchange);
+            exitSpan.getContext().getDestination().getService().withType("messaging").withName("rabbitmq")
+                .getResource().append("rabbitmq/").append(exchange);*/
```

Comment:
Is context required in exit spans? I can't get the producer's instrumentation to work.

Reply:
Without looking too much into the implementation, what do you mean you can't get it to work? What's not working? 
- If you use `log_level=DEBUG`, can you see whether the instrumentation is applied? You should see a `Method match for instrumentation RabbitMQProducerInstrumentation` indicating that the matchers found the method it was looking for
- Can you make sure the producer method is invoked when you expect it to be invoked in the test? If so, you can add a `System.out` and make sure the advice code runs when it is called as well.
- Can you see whether the trace context headers are added to the message? If not, it may be read-only at this point
- Try to print out the message headers on the receiver advice

Reply:
I've updated the `RabbitMQProducerInstrumentationTest` test to check if the trace context headers are added to the message. They are added but the test is failling in the recycling validation.

```
java.lang.RuntimeException: Condition not fulfilled within 1000 ms
Expected :false
Actual   :true
<Click to see difference>


	at co.elastic.apm.agent.MockReporter.awaitUntilAsserted(MockReporter.java:423)
	at co.elastic.apm.agent.MockReporter.awaitUntilAsserted(MockReporter.java:401)
	at co.elastic.apm.agent.MockReporter.assertRecycledAfterDecrementingReferences(MockReporter.java:372)
	at co.elastic.apm.agent.AbstractInstrumentationTest.cleanUp(AbstractInstrumentationTest.java:91)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: org.opentest4j.AssertionFailedError: [should not have any reference left, but has 2 : 'Rabbit-Test Transaction' 00-d7055f0e479e136460efa24fbccdfd1f-23255d7121bbf626-01 (23ee2ccf)] 
Expecting:
 <true>
to be equal to:
 <false>
but was not.
	at jdk.internal.reflect.GeneratedConstructorAccessor18.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at co.elastic.apm.agent.MockReporter.lambda$assertRecycledAfterDecrementingReferences$8(MockReporter.java:384)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.Collections$SynchronizedCollection.forEach(Collections.java:2067)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1085)
	at co.elastic.apm.agent.MockReporter.lambda$assertRecycledAfterDecrementingReferences$9(MockReporter.java:381)
	at co.elastic.apm.agent.MockReporter.awaitUntilAsserted(MockReporter.java:415)
	... 31 more
```

Reply:
That could be because you did not call `span.end()` in `RabbitProducerAdvice#afterBasicPublish()`

Reply:
Yes, seems that it is the error :facepalm: 

Thank you for your patience